src/Routes.ts

```diff
@@ -42,26 +44,100 @@ export interface RouteTreeItem {
 
 export type RouteTree = RouteTreeItem[];
 
-export const resolveRouteStructure = (routes: Routes): RouteTree => {
-  // TODO: Have this build out a route tree e.g.:
-  /*
-    Input: [
-        {label: 'Intro', path: 'intro'},
+/**
+ * Constructs a tree from routing information for use by navigation component
+ * @param routes Routing information
+ * @returns Structured route navigation tree
+ * @example
+ * const input = [
+        {lbel: 'Intro', path: 'intro'},
         {label: 'Push Button', group: 'Inputs/Buttons', path: 'push-button'},
         {label: 'Reset Button', group: 'Inputs/Buttons', path: 'reset-button'}
-    ]
-
-    Output: [
+    ];
+ 
+    const output = [
         { type: 'route', index: 0 },
         { type: 'group', label: 'Inputs', children: [
             { type: 'group', label: 'Buttons', children: [
                 { type: 'route': index: 1 },
                 { type: 'route': index: 2 }
             ]}
         ]}
-    ]
-   
-    */
-  return [];
+    ];
+ */
+export const resolveRouteStructure = (routes: Routes): RouteTree => {
+  const tree: RouteTree = [];
+  const groupRefMap: Record<string, RouteTreeItem> = {};
+
+  const traceGroupRef = (groupName: string): RouteTreeItem[] => {
+    const groupPath = groupName.split(RouteGroupDelimiter);
+    const refs: RouteTreeItem[] = [];
+    for (let i = 0; i < groupPath.length; i++) {
+      const ref =
+        groupRefMap[groupPath.slice(0, i + 1).join(RouteGroupDelimiter)];
+      if (ref) {
+        refs.push(ref);
+      } else {
+        break;
+      }
+    }
+    return refs;
+  };
+
+  const buildGroups = (groupName: string): RouteTreeItem => {
+    const refs = traceGroupRef(groupName);
+    const groupPath = groupName.split(RouteGroupDelimiter).slice(refs.length);
+    let group: RouteTreeItem = {
+      index: -1,
+      type: RouteTreeItemType.GROUP,
+      label: "",
+      children: [],
+    };
+    let parent = refs[refs.length - 1];
+    const groupPrefix = refs.length
+      ? `${groupName
+          .split(RouteGroupDelimiter)
+          .slice(0, refs.length)
+          .join(RouteGroupDelimiter)}${RouteGroupDelimiter}`
+      : "";
+    for (let i = 0; i < groupPath.length; i++) {
+      group = {
+        ...group,
+        label: groupPath[i],
+      };
+      const name = groupPrefix.concat(
+        groupPath.slice(0, i + 1).join(RouteGroupDelimiter)
+      );
+      groupRefMap[name] = group;
+      if (parent) {
+        parent.children?.push(group);
```

Comment:
Wouldn't a parent group always have a `children` property? Rather than runtime optional chaining, perhaps this could be solved by build-time type casting?

```suggestion
        (parent as GroupRouteTreeItem).children.push(group);
```

However, I realized that you don't have separate types for routes vs groups for RouteTreeItem, so doing that would require refactoring the RouteTreeItem type which might not be worth it. This one is nitpicky and not necessary:

```ts
export type RouteTreeItem = TreeItemRoute | TreeItemGroup;

interface TreeItemRoute {
  type: RouteTreeItemType.ROUTE;
  // index of route in Routes if this is a route; default to -1 for group?
  index: number;
}

interface TreeItemGroup {
  type: RouteTreeItemType.GROUP
  // maybe no `index` is needed for groups?
  label: string;
  children: RouteTreeItem[];
}
```